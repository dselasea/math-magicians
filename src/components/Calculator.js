import React, { useState } from 'react';
import PropTypes from 'prop-types';
import calculate from '../logic/calculate';
import Input from './Input';

const Calculator = ({ buttons }) => {
  const [calculation, setCalculation] = useState({ total: 0, next: 0, operation: null });

  const handleCalculation = (e) => {
    const result = calculate(calculation, e.target.name);
    setCalculation(result);
  };

  return (
    <section className="calculator">
      <div className="calculator-title">
        <h1>Let&lsquo;s do some math!</h1>
      </div>
      <div>
        <Input calculation={calculation} />
        <div className="btn-calculator">
          {buttons.map((button) => <button type="button" name={button.btn} onClick={handleCalculation} key={button.id}>{ button.btn }</button>)}
        </div>
      </div>
    </section>
  );
};

Calculator.propTypes = {
  buttons: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number.isRequired,
      btn: PropTypes.string.isRequired,
    }).isRequired,
  ).isRequired,
};

export default Calculator;
