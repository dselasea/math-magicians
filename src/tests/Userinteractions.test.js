import { render, fireEvent } from "@testing-library/react";
import Calculator from "../components/Calculator";

export const buttons = [
  { id: 1, btn: "AC" },
  { id: 2, btn: "+/-" },
  { id: 3, btn: "%" },
  { id: 4, btn: "รท" },
  { id: 5, btn: "7" },
  { id: 6, btn: "8" },
  { id: 7, btn: "9" },
  { id: 8, btn: "x" },
  { id: 9, btn: "4" },
  { id: 10, btn: "5" },
  { id: 11, btn: "6" },
  { id: 12, btn: "-" },
  { id: 13, btn: "1" },
  { id: 14, btn: "2" },
  { id: 15, btn: "3" },
  { id: 16, btn: "+" },
  { id: 17, btn: "0" },
  { id: 18, btn: "." },
  { id: 19, btn: "=" },
];

it("<Calculator /> should do the correct user interactions on pressing the key buttons", () => {
  const calcBody = render(<Calculator buttons={buttons} />);
  fireEvent.click(calcBody.getByText("AC"));
  fireEvent.click(calcBody.getByText("1"));
  fireEvent.click(calcBody.getByText("+"));
  fireEvent.click(calcBody.getByText("1"));
  fireEvent.click(calcBody.getByText("="));

  const displayValue = document.getElementById("input").value.trim();
  expect(calcBody).toMatchSnapshot();
  expect(displayValue).toBe("2");
});
