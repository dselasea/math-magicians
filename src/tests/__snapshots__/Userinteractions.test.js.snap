// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CalcBody /> should render all calculator button 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="calculator"
      >
        <div
          class="calculator-title"
        >
          <h1>
            Let‘s do some math!
          </h1>
        </div>
        <div>
          <div>
            <input
              id="input"
              readonly=""
              type="text"
              value="2  "
            />
          </div>
          <div
            class="btn-calculator"
          >
            <button
              name="AC"
              type="button"
            >
              AC
            </button>
            <button
              name="+/-"
              type="button"
            >
              +/-
            </button>
            <button
              name="%"
              type="button"
            >
              %
            </button>
            <button
              name="÷"
              type="button"
            >
              ÷
            </button>
            <button
              name="7"
              type="button"
            >
              7
            </button>
            <button
              name="8"
              type="button"
            >
              8
            </button>
            <button
              name="9"
              type="button"
            >
              9
            </button>
            <button
              name="x"
              type="button"
            >
              x
            </button>
            <button
              name="4"
              type="button"
            >
              4
            </button>
            <button
              name="5"
              type="button"
            >
              5
            </button>
            <button
              name="6"
              type="button"
            >
              6
            </button>
            <button
              name="-"
              type="button"
            >
              -
            </button>
            <button
              name="1"
              type="button"
            >
              1
            </button>
            <button
              name="2"
              type="button"
            >
              2
            </button>
            <button
              name="3"
              type="button"
            >
              3
            </button>
            <button
              name="+"
              type="button"
            >
              +
            </button>
            <button
              name="0"
              type="button"
            >
              0
            </button>
            <button
              name="."
              type="button"
            >
              .
            </button>
            <button
              name="="
              type="button"
            >
              =
            </button>
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="calculator"
    >
      <div
        class="calculator-title"
      >
        <h1>
          Let‘s do some math!
        </h1>
      </div>
      <div>
        <div>
          <input
            id="input"
            readonly=""
            type="text"
            value="2  "
          />
        </div>
        <div
          class="btn-calculator"
        >
          <button
            name="AC"
            type="button"
          >
            AC
          </button>
          <button
            name="+/-"
            type="button"
          >
            +/-
          </button>
          <button
            name="%"
            type="button"
          >
            %
          </button>
          <button
            name="÷"
            type="button"
          >
            ÷
          </button>
          <button
            name="7"
            type="button"
          >
            7
          </button>
          <button
            name="8"
            type="button"
          >
            8
          </button>
          <button
            name="9"
            type="button"
          >
            9
          </button>
          <button
            name="x"
            type="button"
          >
            x
          </button>
          <button
            name="4"
            type="button"
          >
            4
          </button>
          <button
            name="5"
            type="button"
          >
            5
          </button>
          <button
            name="6"
            type="button"
          >
            6
          </button>
          <button
            name="-"
            type="button"
          >
            -
          </button>
          <button
            name="1"
            type="button"
          >
            1
          </button>
          <button
            name="2"
            type="button"
          >
            2
          </button>
          <button
            name="3"
            type="button"
          >
            3
          </button>
          <button
            name="+"
            type="button"
          >
            +
          </button>
          <button
            name="0"
            type="button"
          >
            0
          </button>
          <button
            name="."
            type="button"
          >
            .
          </button>
          <button
            name="="
            type="button"
          >
            =
          </button>
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Calculator /> should do the correct user interactions on pressing the key buttons 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="calculator"
      >
        <div
          class="calculator-title"
        >
          <h1>
            Let‘s do some math!
          </h1>
        </div>
        <div>
          <div>
            <input
              id="input"
              readonly=""
              type="text"
              value="2  "
            />
          </div>
          <div
            class="btn-calculator"
          >
            <button
              name="AC"
              type="button"
            >
              AC
            </button>
            <button
              name="+/-"
              type="button"
            >
              +/-
            </button>
            <button
              name="%"
              type="button"
            >
              %
            </button>
            <button
              name="÷"
              type="button"
            >
              ÷
            </button>
            <button
              name="7"
              type="button"
            >
              7
            </button>
            <button
              name="8"
              type="button"
            >
              8
            </button>
            <button
              name="9"
              type="button"
            >
              9
            </button>
            <button
              name="x"
              type="button"
            >
              x
            </button>
            <button
              name="4"
              type="button"
            >
              4
            </button>
            <button
              name="5"
              type="button"
            >
              5
            </button>
            <button
              name="6"
              type="button"
            >
              6
            </button>
            <button
              name="-"
              type="button"
            >
              -
            </button>
            <button
              name="1"
              type="button"
            >
              1
            </button>
            <button
              name="2"
              type="button"
            >
              2
            </button>
            <button
              name="3"
              type="button"
            >
              3
            </button>
            <button
              name="+"
              type="button"
            >
              +
            </button>
            <button
              name="0"
              type="button"
            >
              0
            </button>
            <button
              name="."
              type="button"
            >
              .
            </button>
            <button
              name="="
              type="button"
            >
              =
            </button>
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="calculator"
    >
      <div
        class="calculator-title"
      >
        <h1>
          Let‘s do some math!
        </h1>
      </div>
      <div>
        <div>
          <input
            id="input"
            readonly=""
            type="text"
            value="2  "
          />
        </div>
        <div
          class="btn-calculator"
        >
          <button
            name="AC"
            type="button"
          >
            AC
          </button>
          <button
            name="+/-"
            type="button"
          >
            +/-
          </button>
          <button
            name="%"
            type="button"
          >
            %
          </button>
          <button
            name="÷"
            type="button"
          >
            ÷
          </button>
          <button
            name="7"
            type="button"
          >
            7
          </button>
          <button
            name="8"
            type="button"
          >
            8
          </button>
          <button
            name="9"
            type="button"
          >
            9
          </button>
          <button
            name="x"
            type="button"
          >
            x
          </button>
          <button
            name="4"
            type="button"
          >
            4
          </button>
          <button
            name="5"
            type="button"
          >
            5
          </button>
          <button
            name="6"
            type="button"
          >
            6
          </button>
          <button
            name="-"
            type="button"
          >
            -
          </button>
          <button
            name="1"
            type="button"
          >
            1
          </button>
          <button
            name="2"
            type="button"
          >
            2
          </button>
          <button
            name="3"
            type="button"
          >
            3
          </button>
          <button
            name="+"
            type="button"
          >
            +
          </button>
          <button
            name="0"
            type="button"
          >
            0
          </button>
          <button
            name="."
            type="button"
          >
            .
          </button>
          <button
            name="="
            type="button"
          >
            =
          </button>
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Calculator /> should do the correct user interactions on pressing the key buttons all calculator button 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="calculator"
      >
        <div
          class="calculator-title"
        >
          <h1>
            Let‘s do some math!
          </h1>
        </div>
        <div>
          <div>
            <input
              id="input"
              readonly=""
              type="text"
              value="2  "
            />
          </div>
          <div
            class="btn-calculator"
          >
            <button
              name="AC"
              type="button"
            >
              AC
            </button>
            <button
              name="+/-"
              type="button"
            >
              +/-
            </button>
            <button
              name="%"
              type="button"
            >
              %
            </button>
            <button
              name="÷"
              type="button"
            >
              ÷
            </button>
            <button
              name="7"
              type="button"
            >
              7
            </button>
            <button
              name="8"
              type="button"
            >
              8
            </button>
            <button
              name="9"
              type="button"
            >
              9
            </button>
            <button
              name="x"
              type="button"
            >
              x
            </button>
            <button
              name="4"
              type="button"
            >
              4
            </button>
            <button
              name="5"
              type="button"
            >
              5
            </button>
            <button
              name="6"
              type="button"
            >
              6
            </button>
            <button
              name="-"
              type="button"
            >
              -
            </button>
            <button
              name="1"
              type="button"
            >
              1
            </button>
            <button
              name="2"
              type="button"
            >
              2
            </button>
            <button
              name="3"
              type="button"
            >
              3
            </button>
            <button
              name="+"
              type="button"
            >
              +
            </button>
            <button
              name="0"
              type="button"
            >
              0
            </button>
            <button
              name="."
              type="button"
            >
              .
            </button>
            <button
              name="="
              type="button"
            >
              =
            </button>
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="calculator"
    >
      <div
        class="calculator-title"
      >
        <h1>
          Let‘s do some math!
        </h1>
      </div>
      <div>
        <div>
          <input
            id="input"
            readonly=""
            type="text"
            value="2  "
          />
        </div>
        <div
          class="btn-calculator"
        >
          <button
            name="AC"
            type="button"
          >
            AC
          </button>
          <button
            name="+/-"
            type="button"
          >
            +/-
          </button>
          <button
            name="%"
            type="button"
          >
            %
          </button>
          <button
            name="÷"
            type="button"
          >
            ÷
          </button>
          <button
            name="7"
            type="button"
          >
            7
          </button>
          <button
            name="8"
            type="button"
          >
            8
          </button>
          <button
            name="9"
            type="button"
          >
            9
          </button>
          <button
            name="x"
            type="button"
          >
            x
          </button>
          <button
            name="4"
            type="button"
          >
            4
          </button>
          <button
            name="5"
            type="button"
          >
            5
          </button>
          <button
            name="6"
            type="button"
          >
            6
          </button>
          <button
            name="-"
            type="button"
          >
            -
          </button>
          <button
            name="1"
            type="button"
          >
            1
          </button>
          <button
            name="2"
            type="button"
          >
            2
          </button>
          <button
            name="3"
            type="button"
          >
            3
          </button>
          <button
            name="+"
            type="button"
          >
            +
          </button>
          <button
            name="0"
            type="button"
          >
            0
          </button>
          <button
            name="."
            type="button"
          >
            .
          </button>
          <button
            name="="
            type="button"
          >
            =
          </button>
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`<Calculator /> should render all calculator button 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <section
        class="calculator"
      >
        <div
          class="calculator-title"
        >
          <h1>
            Let‘s do some math!
          </h1>
        </div>
        <div>
          <div>
            <input
              id="input"
              readonly=""
              type="text"
              value="2  "
            />
          </div>
          <div
            class="btn-calculator"
          >
            <button
              name="AC"
              type="button"
            >
              AC
            </button>
            <button
              name="+/-"
              type="button"
            >
              +/-
            </button>
            <button
              name="%"
              type="button"
            >
              %
            </button>
            <button
              name="÷"
              type="button"
            >
              ÷
            </button>
            <button
              name="7"
              type="button"
            >
              7
            </button>
            <button
              name="8"
              type="button"
            >
              8
            </button>
            <button
              name="9"
              type="button"
            >
              9
            </button>
            <button
              name="x"
              type="button"
            >
              x
            </button>
            <button
              name="4"
              type="button"
            >
              4
            </button>
            <button
              name="5"
              type="button"
            >
              5
            </button>
            <button
              name="6"
              type="button"
            >
              6
            </button>
            <button
              name="-"
              type="button"
            >
              -
            </button>
            <button
              name="1"
              type="button"
            >
              1
            </button>
            <button
              name="2"
              type="button"
            >
              2
            </button>
            <button
              name="3"
              type="button"
            >
              3
            </button>
            <button
              name="+"
              type="button"
            >
              +
            </button>
            <button
              name="0"
              type="button"
            >
              0
            </button>
            <button
              name="."
              type="button"
            >
              .
            </button>
            <button
              name="="
              type="button"
            >
              =
            </button>
          </div>
        </div>
      </section>
    </div>
  </body>,
  "container": <div>
    <section
      class="calculator"
    >
      <div
        class="calculator-title"
      >
        <h1>
          Let‘s do some math!
        </h1>
      </div>
      <div>
        <div>
          <input
            id="input"
            readonly=""
            type="text"
            value="2  "
          />
        </div>
        <div
          class="btn-calculator"
        >
          <button
            name="AC"
            type="button"
          >
            AC
          </button>
          <button
            name="+/-"
            type="button"
          >
            +/-
          </button>
          <button
            name="%"
            type="button"
          >
            %
          </button>
          <button
            name="÷"
            type="button"
          >
            ÷
          </button>
          <button
            name="7"
            type="button"
          >
            7
          </button>
          <button
            name="8"
            type="button"
          >
            8
          </button>
          <button
            name="9"
            type="button"
          >
            9
          </button>
          <button
            name="x"
            type="button"
          >
            x
          </button>
          <button
            name="4"
            type="button"
          >
            4
          </button>
          <button
            name="5"
            type="button"
          >
            5
          </button>
          <button
            name="6"
            type="button"
          >
            6
          </button>
          <button
            name="-"
            type="button"
          >
            -
          </button>
          <button
            name="1"
            type="button"
          >
            1
          </button>
          <button
            name="2"
            type="button"
          >
            2
          </button>
          <button
            name="3"
            type="button"
          >
            3
          </button>
          <button
            name="+"
            type="button"
          >
            +
          </button>
          <button
            name="0"
            type="button"
          >
            0
          </button>
          <button
            name="."
            type="button"
          >
            .
          </button>
          <button
            name="="
            type="button"
          >
            =
          </button>
        </div>
      </div>
    </section>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
